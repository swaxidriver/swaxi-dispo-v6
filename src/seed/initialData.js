// Deterministic initial seed for Demo (P0-1)
// Generates a fixed 7-day window of shifts with three standard types per day.
// IMPORTANT: Only used when no existing shifts are present (first run / reset).

import { normalizeGeneratedShifts } from '../contexts/ShiftContextCore'

// Base Monday reference (keeps demo stable across weeks)
const SEED_BASE_DATE = '2025-01-06' // Monday
const DAY_COUNT = 7

// Fixed shift templates (name maps to type)
const SHIFT_TEMPLATES = [
  { name: 'Frueh', start: '06:00', end: '14:00' },
  { name: 'Spaet', start: '14:00', end: '22:00' },
  { name: 'Nacht', start: '22:00', end: '06:00' }, // Over-midnight (handled later)
]

function addDays(baseIso, days) {
  const d = new Date(baseIso)
  d.setDate(d.getDate() + days)
  return d.toISOString().slice(0, 10)
}

export function getInitialSeedRaw() {
  const raw = []
  for (let i = 0; i < DAY_COUNT; i++) {
    const dateIso = addDays(SEED_BASE_DATE, i)
    SHIFT_TEMPLATES.forEach(t => {
      raw.push({ date: dateIso, type: t.name, start: t.start, end: t.end })
    })
  }
  return raw
}

export function getInitialSeedShifts() {
  // Normalize into full shift objects (id, status, etc.).
  const normalized = normalizeGeneratedShifts(getInitialSeedRaw())
  // Ensure deterministic ordering (date asc, type asc)
  return normalized.sort((a, b) => (a.date < b.date ? -1 : a.date > b.date ? 1 : a.type.localeCompare(b.type)))
}

export function applyInitialSeedIfEmpty(existing = []) {
  if (existing && existing.length) return existing
  return getInitialSeedShifts()
}

export const __SEED_INTERNALS__ = { SEED_BASE_DATE, DAY_COUNT, SHIFT_TEMPLATES }

// Note: Over-midnight shift (22:00-06:00) currently stored as-is; later story (Zeit/Ãœber-Mitternacht) will refine duration & conflict logic.
