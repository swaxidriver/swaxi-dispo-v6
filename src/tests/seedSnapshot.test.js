import { getInitialSeedShifts } from "../seed/initialData";

describe("Deterministic Seed", () => {
  test("snapshot of initial seed shifts remains stable", () => {
    const shifts = getInitialSeedShifts();
    // Remove volatile props if any (none currently)
    expect(shifts).toMatchSnapshot();
  });
});
