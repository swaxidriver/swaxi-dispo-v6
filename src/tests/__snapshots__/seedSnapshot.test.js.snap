// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Deterministic Seed snapshot of initial seed shifts remains stable 1`] = `
[
  {
    "assignedTo": null,
    "conflicts": [],
    "date": "2025-01-06",
    "end": "14:00",
    "id": "2025-01-06_Frueh",
    "start": "06:00",
    "status": "open",
    "type": "Frueh",
    "workLocation": "office",
  },
  {
    "assignedTo": null,
    "conflicts": [],
    "date": "2025-01-06",
    "end": "06:00",
    "id": "2025-01-06_Nacht",
    "start": "22:00",
    "status": "open",
    "type": "Nacht",
    "workLocation": "office",
  },
  {
    "assignedTo": null,
    "conflicts": [],
    "date": "2025-01-06",
    "end": "22:00",
    "id": "2025-01-06_Spaet",
    "start": "14:00",
    "status": "open",
    "type": "Spaet",
    "workLocation": "office",
  },
  {
    "assignedTo": null,
    "conflicts": [],
    "date": "2025-01-07",
    "end": "14:00",
    "id": "2025-01-07_Frueh",
    "start": "06:00",
    "status": "open",
    "type": "Frueh",
    "workLocation": "office",
  },
  {
    "assignedTo": null,
    "conflicts": [],
    "date": "2025-01-07",
    "end": "06:00",
    "id": "2025-01-07_Nacht",
    "start": "22:00",
    "status": "open",
    "type": "Nacht",
    "workLocation": "office",
  },
  {
    "assignedTo": null,
    "conflicts": [],
    "date": "2025-01-07",
    "end": "22:00",
    "id": "2025-01-07_Spaet",
    "start": "14:00",
    "status": "open",
    "type": "Spaet",
    "workLocation": "office",
  },
  {
    "assignedTo": null,
    "conflicts": [],
    "date": "2025-01-08",
    "end": "14:00",
    "id": "2025-01-08_Frueh",
    "start": "06:00",
    "status": "open",
    "type": "Frueh",
    "workLocation": "office",
  },
  {
    "assignedTo": null,
    "conflicts": [],
    "date": "2025-01-08",
    "end": "06:00",
    "id": "2025-01-08_Nacht",
    "start": "22:00",
    "status": "open",
    "type": "Nacht",
    "workLocation": "office",
  },
  {
    "assignedTo": null,
    "conflicts": [],
    "date": "2025-01-08",
    "end": "22:00",
    "id": "2025-01-08_Spaet",
    "start": "14:00",
    "status": "open",
    "type": "Spaet",
    "workLocation": "office",
  },
  {
    "assignedTo": null,
    "conflicts": [],
    "date": "2025-01-09",
    "end": "14:00",
    "id": "2025-01-09_Frueh",
    "start": "06:00",
    "status": "open",
    "type": "Frueh",
    "workLocation": "office",
  },
  {
    "assignedTo": null,
    "conflicts": [],
    "date": "2025-01-09",
    "end": "06:00",
    "id": "2025-01-09_Nacht",
    "start": "22:00",
    "status": "open",
    "type": "Nacht",
    "workLocation": "office",
  },
  {
    "assignedTo": null,
    "conflicts": [],
    "date": "2025-01-09",
    "end": "22:00",
    "id": "2025-01-09_Spaet",
    "start": "14:00",
    "status": "open",
    "type": "Spaet",
    "workLocation": "office",
  },
  {
    "assignedTo": null,
    "conflicts": [],
    "date": "2025-01-10",
    "end": "14:00",
    "id": "2025-01-10_Frueh",
    "start": "06:00",
    "status": "open",
    "type": "Frueh",
    "workLocation": "office",
  },
  {
    "assignedTo": null,
    "conflicts": [],
    "date": "2025-01-10",
    "end": "06:00",
    "id": "2025-01-10_Nacht",
    "start": "22:00",
    "status": "open",
    "type": "Nacht",
    "workLocation": "office",
  },
  {
    "assignedTo": null,
    "conflicts": [],
    "date": "2025-01-10",
    "end": "22:00",
    "id": "2025-01-10_Spaet",
    "start": "14:00",
    "status": "open",
    "type": "Spaet",
    "workLocation": "office",
  },
  {
    "assignedTo": null,
    "conflicts": [],
    "date": "2025-01-11",
    "end": "14:00",
    "id": "2025-01-11_Frueh",
    "start": "06:00",
    "status": "open",
    "type": "Frueh",
    "workLocation": "office",
  },
  {
    "assignedTo": null,
    "conflicts": [],
    "date": "2025-01-11",
    "end": "06:00",
    "id": "2025-01-11_Nacht",
    "start": "22:00",
    "status": "open",
    "type": "Nacht",
    "workLocation": "office",
  },
  {
    "assignedTo": null,
    "conflicts": [],
    "date": "2025-01-11",
    "end": "22:00",
    "id": "2025-01-11_Spaet",
    "start": "14:00",
    "status": "open",
    "type": "Spaet",
    "workLocation": "office",
  },
  {
    "assignedTo": null,
    "conflicts": [],
    "date": "2025-01-12",
    "end": "14:00",
    "id": "2025-01-12_Frueh",
    "start": "06:00",
    "status": "open",
    "type": "Frueh",
    "workLocation": "office",
  },
  {
    "assignedTo": null,
    "conflicts": [],
    "date": "2025-01-12",
    "end": "06:00",
    "id": "2025-01-12_Nacht",
    "start": "22:00",
    "status": "open",
    "type": "Nacht",
    "workLocation": "office",
  },
  {
    "assignedTo": null,
    "conflicts": [],
    "date": "2025-01-12",
    "end": "22:00",
    "id": "2025-01-12_Spaet",
    "start": "14:00",
    "status": "open",
    "type": "Spaet",
    "workLocation": "office",
  },
]
`;
